core/src/main/scala/org/http4s/headers/Content-Disposition.scala

```diff
@@ -79,13 +83,36 @@ object `Content-Disposition` {
         new Renderable {
           def render(writer: Writer): writer.type = {
             writer.append(v.dispositionType)
-            v.parameters.foreach(p => writer << "; " << p._1 << "=\"" << p._2 << '"')
+            v.parameters.foreach(p => writer << "; " << p._1 << "=" << p._2)
             writer
           }
         },
       parse
     )
 }
 
+sealed trait PtcEncodedString extends Renderable {
+  def render(writer: Writer): writer.type =
+    this match {
+      case AsciiString(value) => writer << s""""$value""""
+      case Utf8String(value) => writer << "UTF-8''" << Uri.encode(value)
+    }
+
+  override def toString = this match {
+    case AsciiString(value) =>
+      value
+    case Utf8String(value) => s"UTF-8''${Uri.encode(value)}"
+  }
+
+}
+final case class AsciiString(s: String) extends PtcEncodedString
+final case class Utf8String(s: String) extends PtcEncodedString
+
+object PtcEncodedString {
+  implicit def fromString(s: String): PtcEncodedString = AsciiString(s)
+}
+
 // see http://tools.ietf.org/html/rfc2183
-final case class `Content-Disposition`(dispositionType: String, parameters: Map[String, String])
+final case class `Content-Disposition`(
+    dispositionType: String,
+    parameters: Map[String, PtcEncodedString])
```

Comment:
Nothing is preventing us from creating `AsciiString("🔥")`, so I'm not sure `PctEncoding` belongs in the public API.  All regular parameters should support an ASCII encoding, and all `*` parameters are supposed to use the charset encoding.  I think we should just store everything as decoded `String`s.

Unfortunately, this simple model does not prevent us from creating invalid key names, or invalid values for keys that aren't the extended tokens.  I don't really see a way to enforce that that results in a pleasant type to use.

The spec does say `filename` and `filename*` are case insensitive, and that duplicate keys aren't supported.  It is silent on whether extension fields are case insensitive, but it's a good bet the answer is yes, so a `Map[CIString, String]` seems appropriate.

Reply:
Very irritatingly, in [RFC7578](https://tools.ietf.org/html/rfc7578), we find the `filename*` MUST NOT be used.  So what's valid depends on whether this is a response header or a multipart part header.  I don't think we can achieve full compliance with all these specs at once without turning this into http4Idris.